Implemented quasiquotations

keddelzz opened the pull request:
## The birth of morganey.meta._

Implemented quasiquotations. Now it's possible to create and interpolate lambda terms with a string-interpolator. Lifting of Scala-values to lambda terms is also possible.

``` scala
val two  = 2
val three = m"3"

val eight = m"( $two $three )"
```

Unquoting of lambda terms allows detection of complex patterns. And unlifting even allows the automatic conversion of lambda terms back to Scala:

``` scala
val term: LambdaTerm = ...

term match {
  case m"( ${a: Int} ${b: Int} )" => m"${math.pow(a, b).toInt}"
  case _ => ???
}
```

This feature was implemented in scope of #25

keddelzz committed:
Add eclipse specific files to gitignore

keddelzz committed:
Add eclipse cache files to gitignore

keddelzz committed:
Split project into subprojects to be able to work with macros.

keddelzz committed:
Implemented special parser, add two interpolators

- The special parser can also parse mangeled names for holes
- Interpolators for quotation of ast nodes: m (= morganey), lc (= lambda
  calculus)

keddelzz committed:
Renamed macro-class

keddelzz committed:
Implemented first version of interpolator

keddelzz committed:
Changed classname after renaming file

keddelzz committed:
Rename decoding functions, Add encoding functions

keddelzz committed:
Add Liftable typeclass for automatic lifting of scala values to lambda terms

keddelzz committed:
Split liftable instances, make them visible in meta._

- Liftable instances are now separate from the Liftable object
- The liftable instances are now visible, if the wildcard import meta._ is
  used

keddelzz committed:
Added liftable instance for Strings

keddelzz committed:
Use liftable instances for conversion from Scala to LambdaTerms

keddelzz committed:
Add simple unquoting, fix building incorrect program

keddelzz committed:
Move sequence function to utils

keddelzz committed:
Add unliftable typeclass for unquoting

keddelzz committed:
Implemented unquoting (without unlifting)

keddelzz committed:
Implemented unlifing of lambda terms, fix Unliftable typeclass

rexim commented:
Looks cool! I'm gonna review this during the day and try to merge the PR before Sunday. Thanks! :)

keddelzz commented:
Should I describe the implemented features in more Detail?

rexim commented:
@keddelzz, The detailization seems to be alright for me. If something is not clear I will ask question in the comments. :) Right now I'm a bit busy. I will take a look into the PR once I'm home. In 5-6 hours or so.

@ForNeVeR, I saw you showed the interest in this PR too. Your comments are appreciated. :)

rexim commented on build.sbt:
```diff
@@ -1,17 +1,61 @@
+  name := "morganey-kernel",
+  libraryDependencies += "org.scala-lang.modules" %% "scala-parser-combinators" % "1.0.4"
+)
 
```
I'm not an expert in macros, but why was it so essential to split the project into subprojects to be able to work with macros? Don't get me wrong I like the idea. :) `meta` works above `kernel` and `kernel` knows nothing about `meta`. This looks right to me. But still what was the motivation to split the project? Was this something related to the technical limitations of macros or just a project structure improvement?

keddelzz replied:
Macro Implementations have to be in a separate projects. They'll have to be compiled before the code, which uses the macro.
Because the macro uses your Parser and the definitions of AST-Nodes, the project structure has to be split. It might be possible to join the definitions of lambda-terms, parser and macros into one subproject, but macros and main-program (for example repl-implementation) have to be separate.

rexim replied:
Got it. Thanks for the explanation!

rexim commented on kernel/src/main/scala/me/rexim/morganey/syntax/LambdaParser.scala:
```diff
@@ -14,7 +14,9 @@ object IntMatcher {
 }
 
-object LambdaParser extends RegexParsers {
+object LambdaParser extends LambdaParser
```
Is this companion object really required here?

keddelzz replied:
Inheritance from your parser was necessary for code reuse (since the interpolator also requires a parser, but with the capability for parsing special names for holes used in interpolation).

The companion object is used by your code in the Main-class. To not break your repl I kept the api which was exposed by your parser.

rexim replied:
Ah, yeah! Right. :) Makes sense. Thanks!

rexim commented on macros/src/main/scala/me/rexim/morganey/meta/QuotationMacro.scala:
```diff
@@ -0,0 +1,161 @@
+import scala.collection.mutable.ArrayBuffer
+import scala.reflect.macros.whitebox.Context
+
+private[meta] class QuotationMacro(val c: Context) {
```
That's a pretty cool class. :) I believe this is where all of the macro magic happens. But it has a lot of imperative logic, which I'd like to cover with any kind of tests that are run on CI. Is it possible to do that quickly? If not we can make it a separate issue and think about it later.

keddelzz replied:
It would be possible to test the result of a macro expansion, but I think it's impossible to run tests while expansion. I'm not quite sure, though. We'd have to try it.

If you want, you could create an issue and assign it to me, I'll add some tests for expansion-results later.

ForNeVeR replied:
I think that the macro library is an ordinary sbt project and it actually can be tested without involving our main project; we could call that our _unit_ tests. At the same time, we can and should also test the results of the macro-expansion, that will be form our _integration_ test layer.

Personally I think that it's preferable to include tests (at least integration ones) in the same PR as the main code. But it'd be okay to write them a bit later. Let @rexim to choose.

keddelzz replied:
The macro project can of course be unit-tested separately, but I think It's impossible to test the macro-code itself (class QuotationMacro). As of now [SIP-5989](https://issues.scala-lang.org/browse/SI-5989) does still exist, which addresses the (unit-)testability of macros.

ForNeVeR replied:
Well, okay then. Our main focus would be on testing the expanded macros anyways.

rexim replied:
Yeah, I'm gonna create a separate issue and we will discuss this later when the time is right.

rexim commented on macros/src/main/scala/me/rexim/morganey/meta/package.scala:
```diff
@@ -0,0 +1,19 @@
+
+  implicit class Quasiquote(ctx: StringContext) {
+    object m extends Quasiquotation
+    object lc extends Quasiquotation
```
What do we need the `lc` interpolator for? It looks like it does the same thing as `m`, no?

keddelzz replied:
Yes, the lc interpolator does completely do the same thing as m. So if you want we could remove it. I thought it would nice to choose between which interpolator to use.

ForNeVeR replied:
Aha, I got that. `m` stands for _Morganey_ and `lc` stands for _Lambda Calculus_. I think it's better to leave only `lc`. Single-letter interpolators is standard library's prerogative I'd say.

keddelzz replied:
I think `m` would be a better name, because it's a reference of the projects name. It would be possible to rename it, to some longer name: `mor` or `mg`, or something like that. If you really wanted it could also be renamed to `morganey`, but I think that'll be too long for the name for an interpolator.

I'm not quite sure why you think single-letter interpolators is standard library's prerogative, after all Scala let's you give them arbitrary names .

keddelzz replied:
Other names i've thought of: `lambda`, `ast` and `tree`. I didn't like all of those ideas, because names were either too long for my taste (for interpolators) or they were not related to the project. 

Ultimately @rexim has to decide, which name he wants.

rexim replied:
I'm not against using `m` or `lc` or both simultaneously. I was just curious why do we have two interpolators that do the same thing. Now I see what they mean. I'm ok with providing them both to the user as available options.

rexim commented:
Currently I actually like what I see. And the quasiquotation seems to be working as expected on my side locally! But I still don't quite understand the `QuotationMacro` class. Mostly because it's almost 3:00am for me! :D

@keddelzz, I'm gonna continue my review tomorrow if you don't mind. I'll try to finish it before the Sunday stream.

Thanks!

keddelzz commented:
Unit tests might have to be rearranged. For now tests for parser and lambda terms are defined in subproject `core`, but they do belong to `kernel`.

rexim commented:
@keddelzz,

> Unit tests might have to be rearranged. For now tests for parser and lambda terms are defined in subproject `core`, but they do belong to `kernel`.

Alright, please do that if it can be done quickly. If not as usual let's just create an issue to not forget about that and move forward. :)

keddelzz commented:
I'll do it as fast as I can!

rexim commented:
@keddelzz I mean, no pressure on that or anything. It's completely optional! I really appreaciate your work! I just want to get your PR merged before 10pm NOVT (in 4 hours) to actually start using the feature on the stream. :) I respect your personal time and if it can't be done before the stream it's totally fine! Really thank you for all of your work!

rexim commented:
Alright, I more or less understand what is happening in the `QuotationMacro` class and I'm ready to merge this PR. Gonna do that right on the stream. :)

keddelzz committed:
Test cleanup

- Move tests to the projects where tested elements were defined in.
- Let sbt run tests of all subprojects and add test-dependencies, so that TestTerms.scala can be found in all tests of all projects
- Add test for sequence function

keddelzz commented:
I've split the tests to their appropriate projects. All builds are passing.

If you want I could write more (integration) tests for macro expansion results later, but this will have to wait a bit. Maybe I could add some in the next week.

rexim commented:
@keddelzz alright! Looks good! Thank you very much! I'm merging this now then.
